File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.py
n = int(input())
d = [[0]+[1]*9 for _ in range(n)]
for i in range(1, n):
    for j in range(10):
        if j == 0:
            d[i][0] = d[i-1][1]
        elif j == 9:
            d[i][9] = d[i-1][8]
        else:
            d[i][j] = (d[i-1][j-1] + d[i-1][j+1]) % int(1e9)
print(sum(d[-1]) % int(1e9))

File: 백준/Silver/10844. 쉬운 계단 수/쉬운 계단 수.java
import java.io.IOException;
import java.util.Arrays;

public class Main {
    static long[][] dp;

    public static void main(String[] args) throws IOException {
        int n = readInt();

        System.out.println(solution(n));
    }

    private static long solution(int n) {
        dp = new long[n][];
        for (int i = 0; i < n; i++) {
            dp[i] = new long[]{0, 1, 1, 1, 1, 1, 1, 1, 1, 1};
        }
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < 10; j++) {
                if (j == 0) {
                    dp[i][0] = dp[i-1][1] % 1000000000;
                } else if (j == 9) {
                    dp[i][9] = dp[i-1][8] % 1000000000;
                } else {
                    dp[i][j] = (dp[i-1][j-1] + dp[i-1][j+1]) % 1000000000;
                }
            }
        }
        return Arrays.stream(dp[n-1]).sum() % 1000000000;
    }

    static int readInt() throws IOException {
        int v, t = 0;
        while ((v = System.in.read()) > 47) {
            t = t * 10 + (v - 48);
        }
        return t;
    }
}